zng/docs/zson.md

```diff
@@ -0,0 +1,629 @@
+* _record_ - an ordered collection of zero or more named values called fields,
+* _array_ - an ordered sequence of zero or more values called elements,
+* _set_ - a set of zero or more unique values called elements,
+* _union_ - a type representing values whose type is any of a specified collection of two or more types,
+* _enum_ - a type representing values taken from a specified collection of one or more uniformly
+typed values, which can be either primitive or complex, and
+* _map_ - a collection of zero or more key/value pairs where the keys are of a
+uniform type called the key type and the values are of a uniform type called
+the value type.
+
+### The Type Type
+
+ZSON also includes first-class types, where a value can be of type `type`.
+The syntax for type values is given below.
+
+### The Null Type
+
+The _null_ type is a primitive type representing only a `null` value.  A `null`
+value can have any type.
+
+Any value in ZSON can take on a null representation.  It is up to an
+implementation to decide how external data structures map into and
+out of values with nulls.  Typically, a null value means either the
+zero value or in the case of record fields, it means the field is optional
+and the value is not present, though these semantics are not explicitly
+defined by ZSON.
+
+### Type Definitions and Contexts
+
+Type definitions embedded in the data sequence associate a name with a type
+so that later values can refer to their type by name instead of explicitly
+enumerating the type of every element.  A collection of mappings from name to
+type is called a "type context."  As bindings are read from a sequence, they
+create the sequence's type context.  The context may be reset at any point in
+the sequence, implying that a new binding must be defined from that point on
+for each unique use of a type.
+
+A type name is either internal or external.  Internal names are used exclusively
+to organize the types in the type context within the data sequence itself and have
+no meaning outside of that data.  External names are visible outside of
+data sequence as named types providing external systems a way to
+discover and refer to types by name where the type is defined within the data
+sequence.  The binding between an external name and its type must be reestablished
+in each type context in which it used.  External names are analogous to the notion
+of "logical types" in Parquet and Avro.
+
+Internal type names are represented by integer names while external names are
+represented by identifiers.  When ZSON data is organized into a sequence
+comprised of two or more subsequences where each subsequence has its own
+type context, the internal names may be "reused" across type contexts to refer
+to different types but external names must have the same type value across
+subsequences.  It is a "type mismatch" error if an external name has different
+type values across subsequences.  It is also a "type mismatch" error if any
+any type binding
```

Comment:
Finish me.

Reply:
Fermat's Last Theorem.  I don't remember what I was about to say.  I will delete until it comes back to me.